### ISSUE TITLE
Unable to load .spydata files containing numpy object arrays 

### ISSUE THREAD

**Issue by nkleinbaer** [spyder-ide/spyder#11754]
### Issue Report Checklist

* [X] Searched the [issues page](https://github.com/spyder-ide/spyder/issues?q=is%3Aissue) for similar reports:
     https://github.com/spyder-ide/spyder/issues/10013 - same issue, but no followup

* [X] Read the relevant sections of the [Spyder Troubleshooting Guide](https://github.com/spyder-ide/spyder/wiki/Troubleshooting-Guide-and-FAQ) and followed its advice
* [X] Reproduced the issue after updating with ``conda update spyder`` (or ``pip``, if not using Anaconda)
* [X ] Could not reproduce inside ``jupyter qtconsole`` (if console-related) (not applicable?)
* [X] Tried basic troubleshooting (if a bug/error)
    * [X] Restarted Spyder
    * [X] Reset preferences with ``spyder --reset``
    * [ ] Reinstalled the latest version of [Anaconda](https://www.anaconda.com/download/)
    * [X ] Tried the other applicable steps from the Troubleshooting Guide
* [X ] Completed the **Problem Description**, **Steps to Reproduce** and **Version** sections below

## Problem Description
When attempting to load .spydata files that contain numpy object arrays, i get the following error: 

![image](https://user-images.githubusercontent.com/47011214/76239281-e892b800-61f6-11ea-8ed2-7293e784da51.png)

I believe the issue is in iofuncs.load_dictionary()
https://github.com/spyder-ide/spyder/blob/9b3e311c3289adeb104d7b60c013e8b34d43b091/external-deps/spyder-kernels/spyder_kernels/utils/iofuncs.py#L415

This [stack exchange post](https://stackoverflow.com/questions/55890813/how-to-fix-object-arrays-cannot-be-loaded-when-allow-pickle-false-for-imdb-loa) provides workaround by modifying the default for np.load() to allow_pickle=True

I'm not sure if this is legitimately a bug or, perhaps a security feature, as per this warning in the numpy documentation [here ](https://docs.scipy.org/doc/numpy/reference/generated/numpy.load.html)

![image](https://user-images.githubusercontent.com/47011214/76244446-ae79e400-61ff-11ea-82df-59ddcd48e21d.png)

### What steps reproduce the problem?

1. Create a numpy array with dtype='object', for example 
![image](https://user-images.githubusercontent.com/47011214/76239361-0d872b00-61f7-11ea-9b81-3b56a68a31db.png)
2. Save environment as foo.spydata using 'Save Data' button above Variable Explorer Pane
3. Restart spyder and attempt to load foo.spydata using 'Import Data' button above Variable Explorer Pane

### What is the expected output? What do you see instead?

Expected out put is that the 'objArray' variable is imported. Instead, error above occurs "Object arrays cannot be loaded when allow_pickle=False"

## Versions

* Spyder version: 4.0.1
* Python version: 3.7.3 (64-bit)
* Qt version: 5.9.6
* PyQt version: 5.9.2
* Operating System name/version: Windows 10

### Dependencies

```
atomicwrites >=1.2.0         :  1.3.0 (OK)
chardet >=2.0.0              :  3.0.4 (OK)
cloudpickle >=0.5.0          :  1.3.0 (OK)
diff_match_patch >=20181111  :  20181111 (OK)
intervaltree                 :  None (OK)
IPython >=4.0                :  7.12.0 (OK)
jedi =0.14.1                 :  0.14.1 (OK)
nbconvert >=4.0              :  5.6.1 (OK)
numpydoc >=0.6.0             :  0.9.2 (OK)
pexpect >=4.4.0              :  4.8.0 (OK)
pickleshare >=0.4            :  0.7.5 (OK)
psutil >=0.3                 :  5.6.7 (OK)
pygments >=2.0               :  2.5.2 (OK)
pylint >=0.25                :  2.4.4 (OK)
pyls >=0.31.2;<0.32.0        :  0.31.7 (OK)
zmq >=17                     :  18.1.1 (OK)
qdarkstyle >=2.7             :  2.8 (OK)
qtawesome >=0.5.7            :  0.6.1 (OK)
qtconsole >=4.6.0            :  4.6.0 (OK)
qtpy >=1.5.0                 :  1.9.0 (OK)
rtree >=0.8.3                :  0.9.3 (OK)
sphinx >=0.6.6               :  2.4.0 (OK)
spyder_kernels >=1.8.1;<2.0.0:  1.8.1 (OK)
watchdog                     :  None (OK)
cython >=0.21                :  0.29.15 (OK)
matplotlib >=2.0.0           :  3.1.3 (OK)
numpy >=1.7                  :  1.18.1 (OK)
pandas >=0.13.1              :  1.0.1 (OK)
scipy >=0.17.0               :  1.4.1 (OK)
sympy >=0.7.3                :  1.5.1 (OK)
```

**Comment by ccordoba12** [comment]
@dalthviz, please take a look at this one asap to see if we can have it for 4.1.0.

**Comment by dalthviz** [comment]
@ccordoba12 as @nkleinbaer said the issue is related with the `allow_pickle` having as default value `False` because of [security reason](https://nvd.nist.gov/vuln/detail/CVE-2019-6446). Should I set `allow_pickle` to `True` in spyder-kernels ?

**Comment by ccordoba12** [comment]
@nkleinbaer, this only happens for object arrays? Or any kind of array?

**Comment by nkleinbaer** [comment]
I've only experienced it w/ object arrays

**Comment by nkleinbaer** [comment]
the numpy docs make it seem like it would only be an issue w/ object arrays
https://docs.scipy.org/doc/numpy/reference/generated/numpy.load.html

**Comment by nkleinbaer** [comment]
@dalthviz, I'm by no means an expert, but I think the security issue with `allow_pickle=True` is moot in this case, since the load and save functions for .spydata files are using pickle anyways? 

> @ccordoba12 as @nkleinbaer said the issue is related with the `allow_pickle` having as default value `False` because of [security reason](https://nvd.nist.gov/vuln/detail/CVE-2019-6446). Should I set `allow_pickle` to `True` in spyder-kernels ?

**Comment by ccordoba12** [comment]
That's not true because someone could put malicious code in a spydata file and send it to another user. So we'll have to inform users about the potential dangers of opening those files every time and they will have to agree on accepting them or not.

**Comment by nkleinbaer** [comment]
Right, but isn't that already a risk when loading ANY spydata file, not only those containing numpy arrays (given that all non-numpy data loaded from a pickle file anyways)? My point was that allowing pickling for numpy in order to fix this bug would not be any less secure than it already is.

If you want to warn before loading anything from a pickle file that sounds like a good idea, but broader issue than fixing this specific bug.

**Comment by nkleinbaer** [comment]
@ccordoba12 I recently noticed this issue is still open and tested it out to confirm it is still happening in a recent release (5.1.5). 

I still think that using`allow_pickle=True` is a fine solution. Above, you mentioned the concern that it would allow someone to put malicious code in a spydata file and send it to another user. But this risk _already exists_, because all objects besides `np.ndarray`s are serialized and deserialized with `pickle`. So allowing it for numpy objects as well doesn't create any _additional_ risk

**Comment by martinclausse** [comment]
This issue isn't resolved for `Spyder version 5.3.2` and `1.23.3`.

**Comment by moreza14** [comment]
Hi, 
I had this problem and the way I got around it was to convert the `numpy` array with `dtype = 'object'` to a list (using `np.tolist()` ) and then save the environment using spyder's interface.  
After using this method spyder loads the .spydata with no errors.  ( I haven't changed the `allow_pickle` at all) 

In my case, I have used  `dtype='object'` to either store something other than a float ( a decimal number ) or a ragged nested sequence. so saving it as a list might be a better option anyways.
